src/radix_tree.cpp

```diff
@@ -0,0 +1,583 @@
+/*
+    Copyright (c) 2018 Contributors as noted in the AUTHORS file
+
+    This file is part of libzmq, the ZeroMQ core engine in C++.
+
+    libzmq is free software; you can redistribute it and/or modify it under
+    the terms of the GNU Lesser General Public License (LGPL) as published
+    by the Free Software Foundation; either version 3 of the License, or
+    (at your option) any later version.
+
+    As a special exception, the Contributors give you permission to link
+    this library with independent modules to produce an executable,
+    regardless of the license terms of these independent modules, and to
+    copy and distribute the resulting executable under terms of your choice,
+    provided that you also meet, for each linked independent module, the
+    terms and conditions of the license of that module. An independent
+    module is a module which is not derived from or based on this library.
+    If you modify this library, you must extend this exception to your
+    version of the library.
+
+    libzmq is distributed in the hope that it will be useful, but WITHOUT
+    ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+    FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
+    License for more details.
+
+    You should have received a copy of the GNU Lesser General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include "precompiled.hpp"
+#include "macros.hpp"
+#include "err.hpp"
+#include "radix_tree.hpp"
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+
+node::node(unsigned char* data)
+    : data_(data)
+{}
+
+uint32_t node::refcount()
+{
+    uint32_t u32;
+    memcpy(&u32, data_, sizeof(u32));
+    return u32;
+}
+
+void node::set_refcount(uint32_t value)
+{
+    memcpy(data_, &value, sizeof(value));
+}
+
+uint32_t node::prefix_length()
+{
+    uint32_t u32;
+    memcpy(&u32, data_ + sizeof(uint32_t), sizeof(u32));
+    return u32;
```

Comment:
Again, unnecessary too complicated.
The whole code in this function can be replaced by:

`return *(reinterpret_cast<uint32_t *>(data_) + 1);`

But, then again, as @sigiesec noted, the cleaner way would be to create a struct where `data_` points to.

Reply:
@bluca 
Yes, you're right, there are more important stuff.
But there's a lot of code like this one and it might be challenging to change it later.
`memcpy()` for small sizes is optimized if compiler implements it as intrinsic function or is very good at inlining.
Even then there's some excess machine code to check for length and alignment of copied block.
So there's **no way** that this could be as optimal as cast.
Another thing. This `memcpy()` is hiding the cast which is still happening. I generally prefer casts to be explicit because they generally mean that you have a mismatched design.

But I think that the real solution is what @sigiesec said, to make a struct and/or union so everything would be much cleaner.

Reply:
@bluca You're right, `memcpy ()` is really optimized out in this case (checked on MSVC 15, release, optimization on), compiler produces the same machine code as if it was a typecast.
But for me it doesn't look readable and the intent of the coder is not clear at first sight.
Anyway, this is minor thing so it can be as it is if you ask me.

Reply:
When the data structure is accessed this way (i.e. without an underlying struct), it should be done with memcpy. While this is probably ok when all of the the header fields are of the same type (it is essentially an array of uint32_t with size 3), this may easily break if some of the header fields is changed to a different type. In that case, using reinterpret_cast may still work and produce the same code on some/most compilers/architectures, but is not portable, since it violates strict aliasing rules. See, e.g. https://en.cppreference.com/w/cpp/language/reinterpret_cast

Reply:
@sigiesec There's no type safety here whatever is used, `memcpy()` or `reinterpret_cast`. `memcpy()` is anyway doing the same casting as `reinterpret_cast` since no types are involved. It's just that `reinterpret_cast` is explicit of what's going on and `memcpy()` is not.
I don't think that aliasing rules are violated in any way since since `uint32_t` is stored in that memory and read again like `uint32_t`, the type is the same. If they are violated then they are violated for both `memcpy()` and `reinterpret_cast`.
What **is** violated is memory alignment of pointers:

```
            // Make room for the new edge. We need to shift the chunk
            // of node pointers one byte to the right. Since resize()
            // increments the edgecount by 1, node_ptrs() tells us the
            // destination address. The chunk of node pointers starts
            // at one byte to the left of this destination.
            //
            // Since the regions can overlap, we use memmove.
            memmove (current_node.node_ptrs (), current_node.node_ptrs () - 1,
                     (current_node.edgecount () - 1) * sizeof (void *));
```
If `std::vector` was used, which is the case for some other implementations of this trie, life would be much easier.

Reply:
My understanding was that using memmove and memcpy would be the safest choice when dealing with raw memory. `node_ptrs` returns an `unsigned char *` so the call to memmove should work as expected, or maybe I'm missing something here?

Reply:
@ssbl 
_My understanding was that using memmove and memcpy would be the safest choice when dealing with raw memory._

You're not dealing with raw memory in a sense that you're processing some data stream from the network or file. The data stored is well known and formed. So the logical choice is to use C++ container objects (`std::vector`, `std::string`) and structs. But since you started it this way then it doesn't matter.
When using `memxxxx()` functions there's also no bounds checking: [SL.con.3: Avoid bounds errors](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#slcon3-avoid-bounds-errors)

Yes, the call to `memmove()` works, I wasn't speaking about that. The memory block where pointers are stored is moved by 1 byte:
  - `current_node.node_ptrs () - 1`

And then data in this block is used as pointers to nodes.
My remark is that the pointers themselves are not aligned in memory to a 4 or 8 byte boundary and are being accessed. This causes high performance penalty on x86 and on some CPUs might not work at all. There's also a case when it doesn't work on x86 when MMX/SSE optimizations are switched on.